---
title: "05 - export"
---

## Running the Export Module

Let's move to the final step of the pipeline: the Export module. This module allows you to create groups of samples, then generate alignments and files formatted for submission to GenBank.

Use the dropdown menu in the top left to navigate to the Export module.

Now let's create a group of samples. Sort the samples table by clicking on the `Taxon` column, then select the first five samples and click the `GROUP` button.

A new window will appear showing some summary information for the selected samples. Type a name for the group, click `Add..`, then click `Create`. You can then click `Close` to return to the sample table.

![Fig. 41](/images/41.png)

The sample table now shows the group assignment in the `Group` column.

![Fig. 42](/images/42.png)

::: {.callout-note}
You can assign new samples to an existing group. Simply select the samples you wish to add, click `GROUP`, then choose the existing group from the dropdown menu.
:::

::: {.callout-note}
Each sample can only belong to one group. MitoPilot will produce a warning message if you attempt to re-assign a sample to a new group.
:::

::: {.callout-note}
Currently, MitoPilot does not allow you to re-export an existing group name.
:::

We can proceed to exporting the data for this group. Click `EXPORT DATA`, which will open a new window. Only samples that have been added to a group can be exported in this manner.

![Fig. 43](/images/43.png)

If you have multiple groups, you can select the appropriate one from the dropdown menu.

MitoPilot allows you to create a custom FASTA header for your samples, pulling data from  columns the CSV file you supplied for the `mapping_fn` argument of the `MitoPilot::new_project` function. This conveniently allows you to include metadata needed for your NCBI GenBank submission. To reference a column, use curly brackets. For example `organism={Taxon}` will autofill values in the "Taxon" column into the FASTA header.

For this test project, we can leave the `Fasta Header Template` at the default value.

The `Generate Group-level PCG alignment summary` will run alignments of all the protein coding genes in your group. For large groups, this can take a while. But it's useful for a final quality control check.

Let's toggle the `Export individual genes` button. This will generate additional FASTA files and GenBank feature tables for each protein coding gene. This can be helpful if you want to use individual genes for phylogenetic analyses or submit them to GenBank. These gene FASTA files have their own custom header template that you can modify.

Click `Export` to generate the final files. It may take a couple of minutes, but as long as the gears are spinning, MitoPilot is still running. 

Once complete, MitoPilot will print the location of the exported files. By default, that location is `YOUR_PROJECT_DIRECTORY/out/export/YOUR_GROUP_NAME`. 

![Fig. 44](/images/44.png)

You can download these files from Hydra or open them from the RStudio server session.

## Exploring the Final Results

MitoPilot will produce three output files. First, the `.html` file contains visualizations of each protein coding gene alignment in your export group. This may be useful to quickly identify samples with poor annotations that need additional manual curation.

![Fig. 45](/images/45.png)

For example, the ATP8 annotation for "SRR22396640" clearly stands out when compared with the other samples in this export group.

![Fig. 46](/images/46.png)

Next are the two files you will need for submission to NCBI GenBank. The `.fasta` file contains the mitogenome assemblies for each sample, following the header template you specified when exporting the data. MitoPilot attempts to adjust the start position of every assembly to the start of the trnF gene.

![Fig. 47](/images/47.png)

Additionally there is the `.tbl` file, a [standardized 5-column, tab-delimited feature table](https://www.ncbi.nlm.nih.gov/genbank/feature_table/) containing information about the annotated genes for each mitogenome.

![Fig. 48](/images/48.png)

With a real dataset, you could use these two files for [submission to NCBI GenBank](https://www.ncbi.nlm.nih.gov/genbank/organelle_submit/).

The export directory also contains a sub-directory `GFFs` with annotations for each sample in [GFF3 format](https://gmod.org/wiki/GFF3). GFF files can be loaded by tools like [Geneious](https://www.geneious.com/) for additional manual inspection prior to submission.

![Fig. 49](/images/49.png)

Lastly, if you selected the `Export individual genes` option, there will be another directory `genes` containing sub-directories for every protein coding gene. In these you will find FASTA files and feature tables for the corresponding gene. There will also be a concatenated FASTA file and feature table containing all protein coding genes, named `GROUP_PCGs.fasta/tbl`.  

![Fig. 50](/images/50.png)
